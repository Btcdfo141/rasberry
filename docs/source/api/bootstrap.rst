.. _bootstrap_module:

:mod:`spencerassistant.bootstrap`
------------------------------

.. automodule:: spencerassistant.bootstrap
    :members:
