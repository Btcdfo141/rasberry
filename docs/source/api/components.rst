:mod:`spencerassistant.components`
===============================

air\_quality
--------------------------------------------

.. automodule:: spencerassistant.components.air_quality
    :members:
    :undoc-members:
    :show-inheritance:

alarm\_control\_panel
--------------------------------------------

.. automodule:: spencerassistant.components.alarm_control_panel
    :members:
    :undoc-members:
    :show-inheritance:

binary\_sensor
--------------------------------------------

.. automodule:: spencerassistant.components.binary_sensor
    :members:
    :undoc-members:
    :show-inheritance:

camera
---------------------------

.. automodule:: spencerassistant.components.camera
    :members:
    :undoc-members:
    :show-inheritance:

calendar
---------------------------

.. automodule:: spencerassistant.components.calendar
    :members:
    :undoc-members:
    :show-inheritance:

climate
---------------------------

.. automodule:: spencerassistant.components.climate
    :members:
    :undoc-members:
    :show-inheritance:

conversation
---------------------------

.. automodule:: spencerassistant.components.conversation
    :members:
    :undoc-members:
    :show-inheritance:

cover
---------------------------

.. automodule:: spencerassistant.components.cover
    :members:
    :undoc-members:
    :show-inheritance:

device\_tracker
---------------------------

.. automodule:: spencerassistant.components.device_tracker
    :members:
    :undoc-members:
    :show-inheritance:

fan
---------------------------

.. automodule:: spencerassistant.components.fan
    :members:
    :undoc-members:
    :show-inheritance:

light
---------------------------

.. automodule:: spencerassistant.components.light
    :members:
    :undoc-members:
    :show-inheritance:

lock
---------------------------

.. automodule:: spencerassistant.components.lock
    :members:
    :undoc-members:
    :show-inheritance:

media\_player
---------------------------

.. automodule:: spencerassistant.components.media_player
    :members:
    :undoc-members:
    :show-inheritance:

notify
---------------------------

.. automodule:: spencerassistant.components.notify
    :members:
    :undoc-members:
    :show-inheritance:

remote
---------------------------

.. automodule:: spencerassistant.components.remote
    :members:
    :undoc-members:
    :show-inheritance:

switch
---------------------------

.. automodule:: spencerassistant.components.switch
    :members:
    :undoc-members:
    :show-inheritance:

sensor
-------------------------------------

.. automodule:: spencerassistant.components.sensor
    :members:
    :undoc-members:
    :show-inheritance:

vacuum
-------------------------------------

.. automodule:: spencerassistant.components.vacuum
    :members:
    :undoc-members:
    :show-inheritance:

water\_heater
-------------------------------------

.. automodule:: spencerassistant.components.water_heater
    :members:
    :undoc-members:
    :show-inheritance:

weather
---------------------------

.. automodule:: spencerassistant.components.weather
    :members:
    :undoc-members:
    :show-inheritance:

webhook
---------------------------

.. automodule:: spencerassistant.components.webhook
    :members:
    :undoc-members:
    :show-inheritance:
