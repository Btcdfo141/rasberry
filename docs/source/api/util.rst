:mod:`spencerassistant.util`
=========================

.. automodule:: spencerassistant.util
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.yaml
-----------------------

.. automodule:: spencerassistant.util.yaml
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.aiohttp
--------------------------

.. automodule:: spencerassistant.util.aiohttp
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.async\_
--------------------------

.. automodule:: spencerassistant.util.async_
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.color
------------------------

.. automodule:: spencerassistant.util.color
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.decorator
----------------------------

.. automodule:: spencerassistant.util.decorator
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.distance
---------------------------

.. automodule:: spencerassistant.util.distance
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.dt
---------------------

.. automodule:: spencerassistant.util.dt
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.json
-----------------------

.. automodule:: spencerassistant.util.json
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.location
---------------------------

.. automodule:: spencerassistant.util.location
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.logging
--------------------------

.. automodule:: spencerassistant.util.logging
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.network
--------------------------

.. automodule:: spencerassistant.util.network
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.package
--------------------------

.. automodule:: spencerassistant.util.package
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.pil
----------------------

.. automodule:: spencerassistant.util.pil
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.pressure
---------------------------

.. automodule:: spencerassistant.util.pressure
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.ssl
----------------------

.. automodule:: spencerassistant.util.ssl
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.temperature
------------------------------

.. automodule:: spencerassistant.util.temperature
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.unit\_system
-------------------------------

.. automodule:: spencerassistant.util.unit_system
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.volume
-------------------------

.. automodule:: spencerassistant.util.volume
   :members:
   :undoc-members:
   :show-inheritance:
