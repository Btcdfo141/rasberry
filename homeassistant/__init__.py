"""Init file for Home Assistant."""


test2
