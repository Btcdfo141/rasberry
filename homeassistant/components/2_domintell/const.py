"""Constants for the Domintell Integration integration."""

DOMAIN = "2_domintell"
