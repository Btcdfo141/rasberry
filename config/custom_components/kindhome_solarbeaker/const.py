from datetime import timedelta

from homeassistant.const import Platform

DOMAIN="kindhome_solarbeaker"
PLATFORMS = [Platform.COVER]

UPDATE_INTERVAL = timedelta(seconds=20)

SERVICE_UUID = "75c276c3-8f97-20bc-a143-b354244886d4"
TITLE="Kindhome Solarbeaker"

DATA_DEVICE = "device"
DATA_COOR = "coordinator"
