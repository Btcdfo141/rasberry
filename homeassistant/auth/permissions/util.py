"""Helpers to deal with permissions."""
from functools import wraps

from typing import (  # noqa: F401
    TYPE_CHECKING, Callable, Dict, List, Optional, Union, cast)

from homeassistant.exceptions import Unauthorized, UnknownUser

from .const import POLICY_CONTROL
from .models import PermissionLookup
from .types import CategoryType, SubCategoryDict, ValueType

if TYPE_CHECKING:
    from homeassistant.core import HomeAssistant, Service, ServiceCall  # noqa

LookupFunc = Callable[[PermissionLookup, SubCategoryDict, str],
                      Optional[ValueType]]
SubCatLookupType = Dict[str, LookupFunc]


def authorized_service_call(hass: 'HomeAssistant', domain: str) -> Callable:
    """Ensure user of a config entry-enabled service call has permission."""
    def decorator(service: 'Service') -> Callable:
        """Decorate."""
        @wraps(service)
        async def check_permissions(call: 'ServiceCall') -> None:
            """Check user permission and raise before call if unauthorized."""
            if not call.context.user_id:
                return

            user = await hass.auth.async_get_user(call.context.user_id)
            if user is None:
                raise UnknownUser(
                    context=call.context,
                    permission=POLICY_CONTROL
                )

            # If the user passes one or more entity IDs, check permissions
            # there; otherwise, check permissions against entities registered
            # to the domain:
            if call.data.get('entity_id'):
                if isinstance(call.data['entity'], str):
                    entities = [call.data['entity_id']]
                else:
                    entities = call.data['entity_id']
            else:
                reg = await hass.helpers.entity_registry.async_get_registry()
                entities = [
                    entity.entity_id for entity in reg.entities.values()
                    if entity.platform == domain
                ]

            for entity_id in entities:
                if user.permissions.check_entity(entity_id, POLICY_CONTROL):
                    return await service(call)

            raise Unauthorized(
                context=call.context,
                permission=POLICY_CONTROL,
            )
        return check_permissions
    return decorator


def lookup_all(perm_lookup: PermissionLookup, lookup_dict: SubCategoryDict,
               object_id: str) -> ValueType:
    """Look up permission for all."""
    # In case of ALL category, lookup_dict IS the schema.
    return cast(ValueType, lookup_dict)


def compile_policy(
        policy: CategoryType, subcategories: SubCatLookupType,
        perm_lookup: PermissionLookup
    ) -> Callable[[str, str], bool]:  # noqa
    """Compile policy into a function that tests policy.
    Subcategories are mapping key -> lookup function, ordered by highest
    priority first.
    """
    # None, False, empty dict
    if not policy:
        def apply_policy_deny_all(entity_id: str, key: str) -> bool:
            """Decline all."""
            return False

        return apply_policy_deny_all

    if policy is True:
        def apply_policy_allow_all(entity_id: str, key: str) -> bool:
            """Approve all."""
            return True

        return apply_policy_allow_all

    assert isinstance(policy, dict)

    funcs = []  # type: List[Callable[[str, str], Union[None, bool]]]

    for key, lookup_func in subcategories.items():
        lookup_value = policy.get(key)

        # If any lookup value is `True`, it will always be positive
        if isinstance(lookup_value, bool):
            return lambda object_id, key: True

        if lookup_value is not None:
            funcs.append(_gen_dict_test_func(
                perm_lookup, lookup_func, lookup_value))

    if len(funcs) == 1:
        func = funcs[0]

        @wraps(func)
        def apply_policy_func(object_id: str, key: str) -> bool:
            """Apply a single policy function."""
            return func(object_id, key) is True

        return apply_policy_func

    def apply_policy_funcs(object_id: str, key: str) -> bool:
        """Apply several policy functions."""
        for func in funcs:
            result = func(object_id, key)
            if result is not None:
                return result
        return False

    return apply_policy_funcs


def _gen_dict_test_func(
        perm_lookup: PermissionLookup,
        lookup_func: LookupFunc,
        lookup_dict: SubCategoryDict
    ) -> Callable[[str, str], Optional[bool]]:  # noqa
    """Generate a lookup function."""
    def test_value(object_id: str, key: str) -> Optional[bool]:
        """Test if permission is allowed based on the keys."""
        schema = lookup_func(
            perm_lookup, lookup_dict, object_id)  # type: ValueType

        if schema is None or isinstance(schema, bool):
            return schema

        assert isinstance(schema, dict)

        return schema.get(key)

    return test_value
