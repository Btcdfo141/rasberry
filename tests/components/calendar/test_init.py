"""The tests for the calendar component."""
from datetime import timedelta, datetime

from homeassistant.bootstrap import async_setup_component
from homeassistant.const import CONF_MAXIMUM
from homeassistant.components.calendar import DOMAIN
from homeassistant.components.websocket_api.const import TYPE_RESULT
import homeassistant.util.dt as dt_util


async def test_events_http_api(hass, hass_client):
    """Test the calendar demo view."""
    await async_setup_component(hass, DOMAIN,
                                {DOMAIN: {'platform': 'demo'}})
    client = await hass_client()
    response = await client.get(
        '/api/calendars/calendar.calendar_2')
    assert response.status == 400
    start = dt_util.now()
    end = start + timedelta(days=1)
    response = await client.get(
        '/api/calendars/calendar.calendar_1?start={}&end={}'.format(
            start.isoformat(), end.isoformat()))
    assert response.status == 200
    events = await response.json()
    assert events[0]['summary'] == 'Future Event'
    assert events[0]['title'] == 'Future Event'


async def test_calendars_http_api(hass, hass_client):
    """Test the calendar demo view."""
    await async_setup_component(hass, DOMAIN,
                                {DOMAIN: {'platform': 'demo'}})
    client = await hass_client()
    response = await client.get('/api/calendars')
    assert response.status == 200
    data = await response.json()
    assert data == [
        {'entity_id': 'calendar.calendar_1', 'name': 'Calendar 1'},
        {'entity_id': 'calendar.calendar_2', 'name': 'Calendar 2'}
    ]


async def test_websocket_get_events_none_found(hass,
                                               hass_ws_client,
                                               calendar_setup):
    """Test websocket endpoint for getting no events found."""
    client = await hass_ws_client(hass)

    await client.send_json({
        'id': 5,
        'entity_id': 'calendar.fake_calendar',
        'type': 'calendar/events',
    })

    msg = await client.receive_json()

    assert msg['id'] == 5
    assert msg['type'] == TYPE_RESULT
    assert not msg['success']
    assert msg['error']['message'] == 'Entity not found'


async def test_websocket_get_all_events(hass,
                                        hass_ws_client,
                                        calendar_setup):
    """Test websocket endpoint for getting all events no filters."""
    client = await hass_ws_client(hass)

    await client.send_json({
        'id': 5,
        'entity_id': 'calendar.real_calendar',
        'type': 'calendar/events',
    })

    msg = await client.receive_json()

    assert msg['id'] == 5
    assert msg['type'] == TYPE_RESULT
    assert msg['success']
    assert len(msg['result']) == 3


async def test_websocket_get_todays_events(hass,
                                           hass_ws_client,
                                           calendar_setup):
    """Test websocket endpoint for getting event filter start/end."""
    client = await hass_ws_client(hass)
    today = datetime.today()

    await client.send_json({
        'id': 5,
        'entity_id': 'calendar.real_calendar',
        'type': 'calendar/events',
        'start': today.isoformat(),
        'end': today.isoformat()
    })

    msg = await client.receive_json()

    assert msg['id'] == 5
    assert msg['type'] == TYPE_RESULT
    assert msg['success']
    assert len(msg['result']) == 1


async def test_websocket_get_max_events(hass,
                                        hass_ws_client,
                                        calendar_setup):
    """Test websocket endpoint for getting event filter start/end."""
    client = await hass_ws_client(hass)
    max_events = 2

    await client.send_json({
        'id': 5,
        'entity_id': 'calendar.real_calendar',
        'type': 'calendar/events',
        CONF_MAXIMUM: max_events
    })

    msg = await client.receive_json()

    assert msg['id'] == 5
    assert msg['type'] == TYPE_RESULT
    assert msg['success']
    assert len(msg['result']) == max_events
