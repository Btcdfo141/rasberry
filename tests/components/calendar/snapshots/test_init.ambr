# serializer version: 1
# name: test_list_events_service[2023-06-22T04:30:00-06:00-2023-06-22T06:30:00-06:00]
  dict({
    'events': list([
      dict({
        'description': 'Future Description',
        'end': '2023-06-22T06:00:00-06:00',
        'location': 'Future Location',
        'start': '2023-06-22T05:00:00-06:00',
        'summary': 'Future Event',
      }),
    ]),
  })
# ---
# name: test_list_events_service[2023-06-22T04:30:00-2023-06-22T06:30:00]
  dict({
    'events': list([
      dict({
        'description': 'Future Description',
        'end': '2023-06-22T06:00:00-06:00',
        'location': 'Future Location',
        'start': '2023-06-22T05:00:00-06:00',
        'summary': 'Future Event',
      }),
    ]),
  })
# ---
# name: test_list_events_service[2023-06-22T10:30:00Z-2023-06-22T12:30:00Z]
  dict({
    'events': list([
      dict({
        'description': 'Future Description',
        'end': '2023-06-22T06:00:00-06:00',
        'location': 'Future Location',
        'start': '2023-06-22T05:00:00-06:00',
        'summary': 'Future Event',
      }),
    ]),
  })
# ---
# name: test_list_events_service_duration[calendar.calendar_1-00:15:00]
  dict({
    'events': list([
    ]),
  })
# ---
# name: test_list_events_service_duration[calendar.calendar_1-01:00:00]
  dict({
    'events': list([
      dict({
        'description': 'Future Description',
        'end': '2023-10-19T05:19:32.771608-07:00',
        'location': 'Future Location',
        'start': '2023-10-19T04:19:32.771608-07:00',
        'summary': 'Future Event',
      }),
    ]),
  })
# ---
# name: test_list_events_service_duration[calendar.calendar_2-00:15:00]
  dict({
    'events': list([
      dict({
        'end': '2023-10-19T04:19:33.058104-07:00',
        'start': '2023-10-19T03:19:33.058104-07:00',
        'summary': 'Current Event',
      }),
    ]),
  })
# ---
